ibis/backends/base/sql/alchemy/__init__.py

```diff
@@ -144,6 +147,44 @@ def _to_geodataframe(df, schema):
             df = gpd.GeoDataFrame(df, geometry=geom_col)
         return df
 
+    def to_pyarrow_batches(
+        self,
+        expr: ir.Expr,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: str = "default",
+        chunk_size: int = 1_000_000,
+    ) -> Iterable[pa.RecordBatch]:
+        self._check_pyarrow_install()
+        import pyarrow as pa
+
+        from ibis.backends.pyarrow.datatypes import sch  # noqa: ignore
+
+        query_ast = self.compiler.to_ast_ensure_limit(
+            expr, limit, params=params
+        )
+        sql = query_ast.compile()
+
+        with self._safe_raw_sql(sql) as cursor:
+            while batch := cursor.fetchmany(chunk_size):
+                struct_array = pa.array(
+                    map(tuple, batch), type=expr.schema().to_pyarrow_struct()
+                )
+                # TODO should we return a RecordBatchReader from this iterator?
+                yield pa.RecordBatch.from_struct_array(struct_array)
+
+    def to_pyarrow(
```

Comment:
Having a default implementation would make it obsolete.

Reply:
How about factoring out these methods to a utility class where we can have sane defaults? Something like:

```py
# should have a better name
class ResultHandler:

    @staticmethod
    def _import_pandas():
        try:
            import pandas
        except ImportError:
            raise ModuleNotFoundError(
                "Exporting to pandas formats requires `pandas` but it is not installed"  # noqa: ignore
            )
        else:
            return pandas

    @staticmethod
    def _import_pyarrow():
        try:
            import pyarrow
        except ImportError:
            raise ModuleNotFoundError(
                "Exporting to arrow formats requires `pyarrow` but it is not installed"  # noqa: ignore
            )
        else:
            return pyarrow

    def to_pylist(self, ...):
        return toolz.concat(self.to_pylist_batches(expr, ...))

    def to_pylist_batches(self, expr, ...):
        ...

    def to_pandas(self, expr, ...):
        pd = self._import_pandas()
        return pd.DataFrame(self.to_pylist(expr, ...))

    def to_pyarrow(self, expr, ...):
        pa = self._import_pyarrow()
        return pa.Table.from_batches(self.to_pyarrow_batches(expr, ...))

    def to_pyarrow_batches(self, expr, ...):
        pa = self._import_pyarrow()
        for batch in self.to_pylist_batches(expr, ...):
            struct_array = pa.array(
                map(tuple, batch), type=expr.schema().to_pyarrow_struct()
            )
            yield pa.RecordBatch.from_struct_array(struct_array)
        

class BaseBackend(abc.ABC, ResultHandler)
    ...

    def execute(self, expr, ...):
        return self.to_pandas(self, expr, ...)
```

Reply:
Tried this out as you suggested and I think I like it.  Not yet re-routing calls to `execute`